bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/DateExpression.java

```diff
@@ -0,0 +1,177 @@
+        // Nothing to do here
+    }
+
+    @Override
+    protected void populateWithSeeds() {
+        // Nothing to do here
+    }
+
+    @Override
+    protected DateExpressionPart parseToken(String token, int position) throws ParseException {
+        return new DateExpressionPart(token);
+    }
+
+    protected class DateExpressionPart extends AbstractExpressionPart {
+
+        private Date theDate;
+
+        public DateExpressionPart(String s) throws ParseException {
+            super(s);
+        }
+
+        public Date getDate() {
+            return theDate;
+        }
+
+        @Override
+        public void parse() throws ParseException {
+            // Unfortunately, the time zone formats available to SimpleDateFormat (Java 6 and earlier) are not ISO 8601
+            // compliant. SimpleDateFormat understands time zone strings like "GMT+01:00" or "+0100", the latter
+            // according to RFC # 822.
+            //
+            // Even if Java 7 added support for time zone descriptors according to ISO 8601, SimpleDateFormat is still
+            // not able to properly parse a complete date string, as it has no support for optional parts.
+            //
+            // Reformatting the input string using regexp is certainly one possibility, but the replacement rules are
+            // not as simple
+            //
+            // Some time zones are not full hours off UTC, so the string does not necessarily end with ":00".
+            // ISO8601 allows only the number of hours to be included in the time zone, so "+01" is equivalent to
+            // "+01:00". ISO8601 allows the usage of "Z" to indicate UTC instead of "+00:00".
+            // The easier solution is to use the data type converter in JAXB, since JAXB must be able to parse
+            // ISO8601 date string according to the XML Schema specification.
+
+            // try {
+            // Calendar cal = DatatypeConverter.parseDateTime(getExpression());
+            // theDate = cal.getTime();
+            // } catch (Exception e) {
+            // throw new ParseException(getPart() + " is not an ISO8601 formatted date", 0);
+            // }
+
+            try {
+                DateFormat df1 = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ssX");
+                df1.setTimeZone(getTimeZone());
+                theDate = df1.parse(getExpression());
+            } catch (Exception e) {
+                try {
+                    DateFormat df2 = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ssXX");
+                    df2.setTimeZone(getTimeZone());
+                    theDate = df2.parse(getExpression());
+                } catch (Exception f) {
+                    try {
+                        DateFormat df3 = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ssXXX");
+                        df3.setTimeZone(getTimeZone());
+                        theDate = df3.parse(getExpression());
+                    } catch (Exception g) {
```

Comment:
Instead of the encapsulated expressions, I would prefer to use another function.
What do you think about e.g. this one:

```java
        /**
         * Try to parse using a set of valid formats.
         *
         * @param formats the date format strings in the order they should be used
         * @return true if parsing succeeded by a format, otherwise false
         */
        private boolean parseFormats(final String[] formats) {
            for (final String format : formats) {
                try {
                    final DateFormat df = new SimpleDateFormat(format);
                    df.setTimeZone(getTimeZone());
                    theDate = df.parse(getExpression());
                    return true;
                } catch (final Exception e) {
                    // Try next one...
                }
            }
            return false;
        }

        @Override
        public void parse() throws ParseException {
            // Unfortunately, the time zone formats available to SimpleDateFormat (Java 6 and earlier) are not ISO 8601
            // compliant. SimpleDateFormat understands time zone strings like "GMT+01:00" or "+0100", the latter
            // according to RFC # 822.
            //
            // Even if Java 7 added support for time zone descriptors according to ISO 8601, SimpleDateFormat is still
            // not able to properly parse a complete date string, as it has no support for optional parts.
            //
            // Reformatting the input string using regexp is certainly one possibility, but the replacement rules are
            // not as simple
            //
            // Some time zones are not full hours off UTC, so the string does not necessarily end with ":00".
            // ISO8601 allows only the number of hours to be included in the time zone, so "+01" is equivalent to
            // "+01:00". ISO8601 allows the usage of "Z" to indicate UTC instead of "+00:00".
            // The easier solution is to use the data type converter in JAXB, since JAXB must be able to parse
            // ISO8601 date string according to the XML Schema specification.

            // try {
            // Calendar cal = DatatypeConverter.parseDateTime(getExpression());
            // theDate = cal.getTime();
            // } catch (Exception e) {
            // throw new ParseException(getPart() + " is not an ISO8601 formatted date", 0);
            // }

            if (!parseFormats(
                    new String[] { "yyyy-MM-dd'T'HH:mm:ssX", "yyyy-MM-dd'T'HH:mm:ssXX", "yyyy-MM-dd'T'HH:mm:ssXXX" })) {
                throw new ParseException(getPart() + " is not an ISO8601 formatted date", 0);
            }
        }
```